src/Microsoft.Diagnostics.Monitoring.StartupHook/MonitorMessageDispatcher/ProfilerMessageSource.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+
+using System.Runtime.InteropServices;
+using System;
+using Microsoft.Diagnostics.Tools.Monitor.Profiler;
+
+namespace Microsoft.Diagnostics.Monitoring.StartupHook.MonitorMessageDispatcher
+{
+    internal sealed class ProfilerMessageSource : IMonitorMessageSource
+    {
+        public event IMonitorMessageSource.MonitorMessageHandler? MonitorMessageEvent;
+
+        public delegate int ProfilerMessageCallback(IpcCommand command, IntPtr nativeBuffer, long bufferSize);
+
+        [DllImport(ProfilerIdentifiers.LibraryRootFileName, CallingConvention = CallingConvention.StdCall, PreserveSig = false)]
+        private static extern void RegisterMonitorMessageCallback(ProfilerMessageCallback callback);
+
+        private static ProfilerMessageSource? s_instance;
+
+        public ProfilerMessageSource()
+        {
+            ProfilerResolver.InitializeResolver<ProfilerMessageSource>();
+            RegisterMonitorMessageCallback(OnProfilerMessage);
+            s_instance = this;
+        }
+
+        private void RaiseMonitorMessage(MonitorMessageArgs e)
+        {
+            MonitorMessageEvent?.Invoke(this, e);
+        }
+
+        private static int OnProfilerMessage(IpcCommand command, IntPtr nativeBuffer, long bufferSize)
```

Comment:
I was under the impression that the Marshaller should be able to handle instance methods as well. If it can't handle them implicitly, you could use Marshal.GetFunctionPointerForDelegate. Just make sure you pin your delegate first.